domain_test.go

```diff
@@ -307,19 +307,243 @@ var domainTestData = []struct {
+			`  </devices>`,
+			`</domain>`,
+		},
+	},
+	{
+		Object: &Domain{
+			Type: "kvm",
+			Name: "test",
+			Devices: &DomainDeviceList{
+				Interfaces: []DomainInterface{
+					DomainInterface{
+						Type: "server",
+						MAC: DomainInterfaceMAC{
+							Address: "52:54:00:39:97:ac",
+						},
+						Model: &DomainInterfaceModel{
+							Type: "virtio",
+						},
+						Source: &DomainInterfaceSource{
+							Address: &[]string{"127.0.0.1"}[0],
+							Port:    &[]uint{1234}[0],
+						},
+					},
+				},
+			},
+		},
+		Expected: []string{
+			`<domain type="kvm">`,
+			`  <name>test</name>`,
+			`  <devices>`,
+			`    <interface type="server">`,
+			`      <mac address="52:54:00:39:97:ac"></mac>`,
+			`      <model type="virtio"></model>`,
+			`      <source address="127.0.0.1" port="1234"></source>`,
+			`    </interface>`,
+			`  </devices>`,
+			`</domain>`,
+		},
+	},
+	{
+		Object: &Domain{
+			Type: "kvm",
+			Name: "test",
+			Devices: &DomainDeviceList{
+				Interfaces: []DomainInterface{
+					DomainInterface{
+						Type: "ethernet",
+						MAC: DomainInterfaceMAC{
+							Address: "52:54:00:39:97:ac",
+						},
+						Model: &DomainInterfaceModel{
+							Type: "virtio",
+						},
+						Script: &DomainInterfaceScript{
+							Path: "/etc/qemu-ifup",
+						},
+					},
+				},
+			},
+		},
+		Expected: []string{
+			`<domain type="kvm">`,
+			`  <name>test</name>`,
+			`  <devices>`,
+			`    <interface type="ethernet">`,
+			`      <mac address="52:54:00:39:97:ac"></mac>`,
+			`      <model type="virtio"></model>`,
+			`      <script path="/etc/qemu-ifup"></script>`,
+			`    </interface>`,
+			`  </devices>`,
+			`</domain>`,
+		},
+	},
+	{
+		Object: &Domain{
+			Type: "kvm",
+			Name: "test",
+			Devices: &DomainDeviceList{
+				Interfaces: []DomainInterface{
+					DomainInterface{
+						Type: "vhostuser",
+						MAC: DomainInterfaceMAC{
+							Address: "52:54:00:39:97:ac",
+						},
+						Model: &DomainInterfaceModel{
+							Type: "virtio",
+						},
+						Source: &DomainInterfaceSource{
+							Type: &[]string{"unix"}[0],
+							Path: &[]string{"/tmp/vhost0.sock"}[0],
+							Mode: &[]string{"server"}[0],
+						},
+					},
+				},
+			},
+		},
+		Expected: []string{
+			`<domain type="kvm">`,
+			`  <name>test</name>`,
+			`  <devices>`,
+			`    <interface type="vhostuser">`,
+			`      <mac address="52:54:00:39:97:ac"></mac>`,
+			`      <model type="virtio"></model>`,
+			`      <source type="unix" path="/tmp/vhost0.sock" mode="server"></source>`,
+			`    </interface>`,
+			`  </devices>`,
+			`</domain>`,
+		},
+	},
 }
 
 func TestDomain(t *testing.T) {
 	for _, test := range domainTestData {
 		doc, err := test.Object.Marshal()
 		if err != nil {
-			t.Fatal(err)
+			t.Error(err)
 		}
 
 		expect := strings.Join(test.Expected, "\n")
 
 		if doc != expect {
-			t.Fatal("Bad xml:\n", string(doc), "\n does not match\n", expect, "\n")
+			t.Error("Bad xml:\n", string(doc), "\n does not match\n", expect, "\n")
```

Comment:
These changes are unrelated to the rest of the commit - please put them in a separate commit and why why they're needed